### PR TITLE
feat(watchlist): implement Plex watchlist GraphQL API support

### LINKED ISSUE #1378: [Feature Request] Fetch Plex user watchlists without needing their Plex user token
### Description

Currently, Jellyseerr can only fetch watchlists for Plex users who have signed in to Jellyseerr through Plex OAuth and given Jellyseerr their Plex user token. If an imported Plex user has no token (i.e. hasn't signed in manually), they are ignored and their watchlist cannot be synced:
https://github.com/fallenbagel/jellyseerr/blob/525a538f34dfec3bd41974f6b37c1077242f1987/server/lib/watchlistsync.ts#L20-L26
https://github.com/fallenbagel/jellyseerr/blob/525a538f34dfec3bd41974f6b37c1077242f1987/server/lib/watchlistsync.ts#L33-L39

This is undesirable because it requires every user to manually sign in to Jellyseerr in order to be able to use their Plex watchlist to make requests, as opposed to seamlessly integrating with Plex without any user intervention needed.

### Desired Behavior

I'd like to allow my Plex users to seamlessly make requests through their Plex watchlists without ever needing to log in to Jellyseerr. This can be achieved by querying Plex's GraphQL  API (as used by [Watchlistarr](https://github.com/nylonee/watchlistarr)) which would allow Jellyseerr to fetch watchlists without needing a user's Plex token. It uses the server admin's Plex token to make GraphQL requests to fetch the admin's Plex friends' watchlists.

Requirements:
- the server admin's Plex account is Plex friends with the target Plex account
- the target Plex account has their watchlist visibility settings configured to allow friends to view their watchlist.

The relevant implementation is here: https://github.com/nylonee/watchlistarr/blob/dc9ef8521d76428127cf112ad571d2eacb65e074/src/main/scala/plex/PlexUtils.scala#L141

Watchlistarr also implements creating watchlist RSS feeds for Plex friends (which seems like great functionality to pull here) but the feature is listed as "currently broken" in the documentation with no additional info as to _how_ it's broken. Not sure if it actually works or not.

### Additional Context

I realize this is a Plex-focused feature request in a Jellyfin/Emby-focused fork, but Overseerr development seems to be completely dead at this point so I figured this was the only place for it. I'd love to try and get a PR up for this myself, but I have little to no experience with Typescript and this feels a little outside my wheelhouse. I'd love some feedback on this!

### Code of Conduct

- [x] I agree to follow Jellyseerr's Code of Conduct

### PR DESCRIPTION
**Pull request by xNul**
## Description

This PR builds off of #1426 and aims to close #1378.

Currently, Seerr's Plex watchlist sync functionality requires each Plex user's individual user token. This is a burden on the user because it means they have to get the login credentials for every Plex user's account or have the user access Seerr and login themselves. For situations where obtaining a user's token isn't possible or there is no desire to give access to Seerr's web interface, Seerr hasn't been able to do a watchlist sync.

However, this is possible to do with Plex's GraphQL API. The only requirement to access a user's watchlist via the GraphQL API is that the user's watchlist is accessible to the user token which is making the API call. For example, if a user makes their watchlist public, it can be synced. If a user is friends with the user token which is making the API call, it can be synced.

Since the owner of a Seerr instance is the owner of the Plex media server, the Plex media server owner can add their users as friends (an easier task, if they aren't already friends) and have access to all of the watchlists of the media server. This means, there is no need to have each Plex user login to Seerr in order to provide Plex watchlist sync functionality.

Originally, PR #1426 was [closed by the author](https://github.com/seerr-team/seerr/pull/1426#issuecomment-2741966172) with the expectation Plex was about to provide a public API which would officially solve this issue. It has been over a year now and as far as I can tell, no such API has been released so I have brought the PR back.

Progress Checklist
- [x] Rebased and modified for the latest commit
- [x] Removed legacy watchlist sync API
- [x] Decoupled `getWatchlist` from API token
- [x] Robust way to fetch and store UUIDs
- [x] Restored watchlist caching functionality
- [x] Database migrations for the new UUID column
- [x] Handles database migration edge cases where UUID values may be `null`
- [x] Fixed tests
- [x] Added an access check when enabling watchlist sync for a user
- [ ] Handle pagination differences
- [ ] Simplify SQLite migration code?
- [ ] Better `getWatchlist` error handling?

Important Changes
- A new `plexUuid` column has been added to the `User` entity/table.
- The legacy watchlist sync API has been removed because it was redundant. All previous functionality has been retained.
- A new Plex API `/api/home/users` is used to provide all the Plex UUIDs and replaces the `/api/users` Plex API call during `/import-from-plex` Plex user import.

## The Pagination Issue

Pagination in Seerr is built around offsets, but the GraphQL watchlist API uses cursors. [More info here](https://dev.to/jackmarchant/offset-and-cursor-pagination-explained-b89). We need a way to bridge the gap and I'm not sure which way to go. I have marked the respective areas where pagination needs to be fixed with `TODO` comments.

The page is set via the `page` query parameter
https://github.com/seerr-team/seerr/blob/dfde4d34e69732935c0ffa72da4d1f15469b6774/server/routes/user/index.ts#L951
which can be set 1-n at any point in time. With offset-based pagination, that's a simple operation that only returns `page_size` results for a given `page`, but cursor-based pagination would require consuming `page_size*page` results to find the results for a given `page`. Page caching should help here though.

Edit: The cursor for each watchlist item is simply when the item was added in unix epoch time encoded in base64. We can do a lot with this knowledge.

Edit 2: I have reverse engineered the GraphQL fields for `GetWatchlistHub`. There may still be more, but I don't think there are any left that are related to pagination.

<details>
<summary>Click to expand GetWatchlistHub example query</summary>

```graphql
fragment imageItems on MetadataItemImageSet {
  coverArt
  coverPoster
  thumbnail
  art
  __typename
}

fragment userStateItems on UserState {
  viewCount
  viewedLeafCount
  watchlistedAt
  __typename
}

fragment parentFields on MetadataItem {
  id
  title
  type
  index
  key
  publishedAt
  __typename

  images {
    ...imageItems
  }

  userState {
    ...userStateItems
  }
}

fragment itemFields on MetadataItem {
  id
  title
  type
  year
  guid
  key
  slug

  duration
  index
  publishedAt
  originallyAvailableAt
  publicPagesURL

  leafCount
  childCount

  tagline
  imdbRatingCount

  __typename

  parent {
    ...parentFields
  }

  grandparent {
    ...parentFields
  }

  userState @skip(if: $skipUserState) {
    ...userStateItems
  }

  images {
    ...imageItems
  }
}

# `first` and `last` can't be used at the same time. One must be removed.
query GetWatchlistHub(
  $uuid: ID = ""
  $first: PaginationInt!
  $after: String
  $last: PaginationInt!
  $before: String
  $skipUserState: Boolean = false
) {
  user(id: $uuid) {
    watchlist(
      first: $first
      after: $after
      last: $last
      before: $before
    ) {
      nodes {
        ...itemFields
      }

      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
}
```

</details>

## How Has This Been Tested?

- Passes `pnpm test`
- Manual testing of each edge case
  - Updating an old database to a new UUID database
  - Plex watchlists with no items
  - Plex watchlists with items
  - Plex watchlists without access
  - Plex watchlists with access
  - Plex watchlists with access -> set to private -> without access auto-detection and configuration
  - Watchlist access check trigger before migration UUID column population
  - Login and watchlist sync with another Plex user
  - Running from scratch with no existing config or database
- I have been running it myself for a small set of users

## Screenshots / Logs (if applicable)

Example attempt to enable plex watchlist sync without access to it

image

## Checklist:

- [x] I have read and followed the contribution [guidelines](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md).
- [x] Disclosed any use of AI (see our [policy](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md#ai-assistance-notice))
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] Successful build `pnpm build`
- [x] Translation keys `pnpm i18n:extract`
- [x] Database migration (if required)

**AI Disclosure:** The equivalent of Googling. AI was not used to write any of this code.

## Summary by CodeRabbit

* **New Features**
  * Added Plex home users support and migrated watchlist sync to UUID-based identification with cursor-style pagination; backend now persists Plex UUIDs and seeds test accounts with UUIDs.

* **Bug Fixes**
  * Improved handling when a Plex watchlist is inaccessible: clearer error, user-facing toast, and automatic disabling of watchlist sync when access fails.

* **Documentation**
  * Updated Plex Watchlist Auto Request docs to clarify access prerequisites and failure guidance.

[Review Change Stack](https://app.coderabbit.ai/change-stack/seerr-team/seerr/pull/3008)

### CODE REVIEW

File: server/lib/watchlistsync.ts [hunk 1]
```diff
@@ -18,22 +19,64 @@ class WatchlistSync {
...
+        if (account.email) {
+          const user = await userRepository
+            .createQueryBuilder('user')
+            .where('user.plexId = :id', { id: account.id })
+            .getOne();
+
+          if (user) {
+            user.plexUuid = account.uuid;
+            await userRepository.save(user);
+          }
```
[the remark this answers is not in the document]
Comment (xNul) [reply server/lib/watchlistsync.ts]: I took this from
https://github.com/seerr-team/seerr/blob/dfde4d34e69732935c0ffa72da4d1f15469b6774/server/routes/user/index.ts#L673-L698
The email check is done here too. I believe that's because only Plex users will have emails so this is a check for if they are a Plex user.

File: server/routes/user/usersettings.ts [hunk 2]
```diff
@@ -119,6 +131,41 @@ userSettingsRoutes.post<
...
+        const plexHomeUsersResponse = await plexTvApi.getHomeUsers();
+        for (const rawUser of plexHomeUsersResponse.MediaContainer.User) {
+          const account = rawUser.$;
+
+          if (account.email && Number(account.id) === user.plexId) {
+            user.plexUuid = account.uuid;
+            await userRepository.save(user);
+            break;
+          }
+        }
```
[the remark this answers is not in the document]
Comment (xNul) [reply server/routes/user/usersettings.ts:158]: https://github.com/seerr-team/seerr/pull/3008#discussion_r3214372496

File: server/api/plextv.ts [hunk 3]
```diff
@@ -268,33 +291,63 @@ class PlexTvAPI extends ExternalAPI {
...
+    {
+      first = 20,
+      after = undefined,
+    }: { first?: number; after?: number | undefined } = {}
+  ): Promise<{
+    first: number;
+    after: number | undefined;
+    endCursor: string | undefined;
     totalSize: number;
     items: PlexWatchlistItem[];
```
[the remark this answers is not in the document]
Comment (xNul) [reply server/api/plextv.ts:317]: Known issue and to be done as stated in the PR description.

File: server/routes/user/usersettings.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (xNul) [reply server/routes/user/usersettings.ts:158]: Oh that makes sense. Fixing it now.

**Comment by xNul** [comment]
For those 8 outside diff range + nitpick comments:
1. Fixed
2. See "The Pagination Issue" in the description
3. Fixed
4. I guess I could add an `accessDenied` boolean return value? Pretty janky that way too.
5. Fixed
6. Fixed
7. This is just #⁣4
8. Fixed

Which just leaves pagination and maybe #⁣4.

File: server/api/plextv.ts [hunk 4]
```diff
@@ -268,35 +291,65 @@
...
-    offset = 0,
-    size = 20,
-  }: { offset?: number; size?: number } = {}): Promise<{
-    offset: number;
-    size: number;
+  public async getHomeUsers(): Promise<HomeUsersResponse> {
+    const response = await this.axios.get('/api/home/users', {
+      transformResponse: [],
+      responseType: 'text',
+    });
```
[the remark this answers is not in the document]
Comment (xNul) [reply server/api/plextv.ts:298]: I don't see how the request host depends on a user-provided value... The host is hard-coded at
https://github.com/seerr-team/seerr/blob/dfde4d34e69732935c0ffa72da4d1f15469b6774/server/api/plextv.ts#L143

[the remark this answers is not in the document]
Comment (0xSysR3ll) [reply server/api/plextv.ts:298]: Dw that's just a false positive. I dismissed it ;)

[the remark this answers is not in the document]
Comment (xNul) [reply server/api/plextv.ts:298]: @0xSysR3ll looks like it's still failing the code scanner
image
Not sure why since this is the only problem it raised and you've dismissed it.

[the remark this answers is not in the document]
Comment (0xSysR3ll) [reply server/api/plextv.ts:298]: There was another one, should be good now.
